compositor_render/src/renderer/texture.rs

```diff
@@ -25,66 +29,90 @@ pub type YUVTextures = yuv::YUVTextures;
 
 pub type Texture = base::Texture;
 
-pub struct InputTexture {
+pub struct InputTextureState {
     textures: YUVTextures,
     bind_group: wgpu::BindGroup,
-    resolution: Resolution,
-    init_color: RGBColor,
 }
 
-impl InputTexture {
-    pub fn new(ctx: &WgpuCtx, init_color: Option<RGBColor>) -> Self {
-        let resolution = Resolution {
-            width: 2,
-            height: 2,
-        };
-        let textures = YUVTextures::new(ctx, resolution);
-        let bind_group = textures.new_bind_group(ctx, ctx.format.yuv_layout());
+impl InputTextureState {
+    pub fn yuv_textures(&self) -> &YUVTextures {
+        &self.textures
+    }
 
-        let init_color = init_color.unwrap_or(RGBColor(0, 0, 0));
-        let (y, u, v) = init_color.to_yuv();
-        ctx.utils.fill_r8_with_value(ctx, textures.plane(0), y);
-        ctx.utils.fill_r8_with_value(ctx, textures.plane(1), u);
-        ctx.utils.fill_r8_with_value(ctx, textures.plane(2), v);
+    pub fn bind_group(&self) -> &wgpu::BindGroup {
+        &self.bind_group
+    }
 
-        Self {
-            textures,
-            bind_group,
-            resolution,
-            init_color,
-        }
+    pub fn resolution(&self) -> Resolution {
+        self.textures.resolution
+    }
+}
+
+pub struct InputTexture(OptionalState<InputTextureState>);
+
+impl InputTexture {
+    pub fn new() -> Self {
+        Self(OptionalState::new())
+    }
+
+    pub fn clear(&mut self) {
+        self.0.clear()
     }
 
     pub fn upload(&mut self, ctx: &WgpuCtx, frame: Frame) {
-        if frame.resolution != self.resolution {
-            self.textures = YUVTextures::new(ctx, frame.resolution);
-            self.bind_group = self.textures.new_bind_group(ctx, ctx.format.yuv_layout());
-            self.resolution = frame.resolution;
-
-            let (y, u, v) = self.init_color.to_yuv();
-            ctx.utils.fill_r8_with_value(ctx, self.textures.plane(0), y);
-            ctx.utils.fill_r8_with_value(ctx, self.textures.plane(1), u);
-            ctx.utils.fill_r8_with_value(ctx, self.textures.plane(2), v);
+        let state = self.ensure_size(ctx, frame.resolution);
+        state.textures.upload(ctx, &frame.data)
+    }
+
+    fn ensure_size<'a>(
+        &'a mut self,
+        ctx: &WgpuCtx,
+        new_resolution: Resolution,
+    ) -> &'a InputTextureState {
+        fn new_state(ctx: &WgpuCtx, new_resolution: Resolution) -> InputTextureState {
+            let textures = YUVTextures::new(ctx, new_resolution);
+            let bind_group = textures.new_bind_group(ctx, ctx.format.yuv_layout());
+            InputTextureState {
+                textures,
+                bind_group,
+            }
         }
-        self.textures.upload(ctx, &frame.data)
+
+        self.0 = match self.0.replace(OptionalState::None) {
```

Comment:
flagging this so it does not get missed during review. I couldn't find an approach where I could avoid this.
I don't need that for NodeTexture because I can clone that state.